01 - JavaScript Drum Kit/index-START.html

```diff
@@ -1,66 +1,122 @@
-    <div data-key="74" class="key">
-      <kbd>J</kbd>
-      <span class="sound">snare</span>
-    </div>
-    <div data-key="75" class="key">
-      <kbd>K</kbd>
-      <span class="sound">tom</span>
-    </div>
-    <div data-key="76" class="key">
-      <kbd>L</kbd>
-      <span class="sound">tink</span>
-    </div>
-  </div>
+    <script>
+      // selectors
+      // select all keys
+      const keys = document.querySelectorAll('.key');
+
+      // callbacks
+      // listen to keyDowns
+      const playAndAddClass = (evt) => {
+        //console.log(evt.keyCode);
+        // select only the audio tag = to the key you press on the keyboard
+        // use attribute selector for this 'audio[data-key = evt.keyCode]'
+        const audio = document.querySelector(
+          `audio[data-key="${evt.keyCode}"]`
+        );
+
+        // select only the div tag = to the key you press on the keyboard
+        const key = document.querySelector(`div[data-key="${evt.keyCode}"]`);
+
+        // run the function only if the key = one of our keyCodes
+        if (!audio) return;
+
+        // rewind every time you click, basically don't wait for the audio to finish
+        audio.currentTime = 0;
+
+        // play the audio if any audio is typed
+        audio.play();
+
+        // add class to style key when keyDown
+        key.classList.add('playing');
+      };
+
+      // I first thought I could use a keyUp event to remove the playing class
+      // This does not work if you have a long transition duration
+      // You basically do not see the transition at all because the keyUp is faster
+      /*
+      const keyUpListener = (evt) => {
+        // select only the div tag = to the key you press on the keyboard
+        const key = document.querySelector(`div[data-key="${evt.keyCode}"]`);
 
-  <audio data-key="65" src="sounds/clap.wav"></audio>
-  <audio data-key="83" src="sounds/hihat.wav"></audio>
-  <audio data-key="68" src="sounds/kick.wav"></audio>
-  <audio data-key="70" src="sounds/openhat.wav"></audio>
-  <audio data-key="71" src="sounds/boom.wav"></audio>
-  <audio data-key="72" src="sounds/ride.wav"></audio>
-  <audio data-key="74" src="sounds/snare.wav"></audio>
-  <audio data-key="75" src="sounds/tom.wav"></audio>
-  <audio data-key="76" src="sounds/tink.wav"></audio>
+        // run the function only if the key = one of our keyCodes
+        if (!key) return;
 
-<script>
+        // remove class to style key when keyDown
+        key.classList.remove('playing');
+      };
+      */
 
-</script>
+      // listeners
+      // listen for keyDowns to add playing class and play sound
+      window.addEventListener('keydown', playAndAddClass);
 
+      // keyUp does not show transition if transition duration is long
+      // window.addEventListener('keyup', keyUpListener);
 
-</body>
+      // listen for transitionEnd to remove playing class
+      keys.forEach((key) =>
+        key.addEventListener('transitionend', (evt) => {
```

Comment:
can you discuss when it is cool to use the `transitionend` event? maybe have a look at `https://caniuse.com/`

Reply:
super cool website Tom! thank you for the tip! I just realised that the transitionEnd event is fully compatible only on Firefox and Safari. Most other browsers support only the event when happening on an HTMLElement but not on Window and Document. In general this event fits the needs of our small DrumKit because we want to see the transition regardless of what its duration is (long or short) or the speed the user types. The keyUp event I used at the beginning was dependent on the speed the user was typing on the keys, leaving the DrumKit with no control on the transition itself.